### PR TITLE
Overhaul of the examples folder

### PR DESCRIPTION
**Pull request by ernestum**
To resolve issue #383 this MR introduces a couple of new example files.

### CODE REVIEW

File: examples/utils.py
Comment (ernestum) [inline examples/utils.py:1]: My gut feeling says, that this function should already be in some gym utils, but I could not find it. Does anybody have a hint for me?

Comment (AdamGleave) [reply examples/utils.py:1]: You could do this using `InteractiveRender` from https://github.com/HumanCompatibleAI/imitation/blob/master/src/imitation/scripts/eval_policy.py#L22 (move it from script to util), `rollout.generate_trajectories` and `rollout.rollout_stats`. Would need to wrap the env into a `VecEnv` first. Not obviously cleaner given the number of utilities needed, but introduces the reader to some relevant code at least.

Comment (ernestum) [reply examples/utils.py]: I went with `stable_baselines3.common.evaluation.evaluate_policy` for now. There is a `render` flag, that we can just set to `True`.

File: src/imitation/data/rollout.py
Comment (AdamGleave) [inline src/imitation/data/rollout.py:600]: I think `rollout_and_save` is only used in `train_rl`, so I'm tempted to just remove this function given it's basically a one-liner anyway

**Review by AdamGleave — COMMENTED** [review]
Thanks, this is shaping up nicely!

I've left some minor comments. A few higher-level notes:
  - We really want `examples/quickstart.py` to actually be runnable, so having a dependency on `pytest_cache` is not ideal. 
  - I'm worried the notebooks will get broken over time and we won't realize. Can we add some smoke tests to make sure they run? I think `pytest-notebook` configured to [ignore the output](https://pytest-notebook.readthedocs.io/en/latest/user_guide/tutorial_config.html) would do this. (Or we could try to make the output actually deterministic by seeding, but that sounds tricky.)

File: examples/4_train_airl.ipynb
Comment (ernestum) [reply examples/4_train_airl.ipynb]: No this was just an accident. I removed it.

File: src/imitation/data/rollout.py
Comment (ernestum) [reply src/imitation/data/rollout.py]: Good point. I removed it.

File: examples/1_train_bc.ipynb
Comment (ernestum) [reply examples/1_train_bc.ipynb]: Good point! I adapted the text.

**Comment by ernestum** [comment]
I am a bit stuck with the MCE example. I figured out how to make it execute in some sensible way but the results are garbage. Maybe I need a different RL algorithm for the learner? I just took PPO with the same hyper-parameters as for CartPole which is probably not the smartest solution. Do you have any hints for me?

File: examples/quickstart.py
Comment (ernestum) [reply examples/quickstart.py:1]: For me the purpose is to have a minimal example, that fits (more or less, if you ignore the imports) on a single screen.
It also allows you to quickly launch something without having to spin up a jupyter notebook (in case that adds friction).
However, I see your point that the difference to the `train_bc.ipynb` is minimal right now. The only thing I really like is that in the python script we can set render=True to actually show what is going on.

**Comment by ernestum** [comment]
> * I'm worried the notebooks will get broken over time and we won't realize. Can we add some smoke tests to make sure they run? I think `pytest-notebook` configured to [ignore the output](https://pytest-notebook.readthedocs.io/en/latest/user_guide/tutorial_config.html) would do this. (Or we could try to make the output actually deterministic by seeding, but that sounds tricky.)

There are actually already a `test_run_example_notebooks` test case in `test_examples.py`. 
It just runs the notebooks to verify that there are no exceptions. 
I think this is already quite a good smoke test.
However, I disabled the test for now using `pytest.mark.skip` since some of the notebooks take quite long.
I think there are two options here:
- We tweak the parameters in the notebooks so they finish quickly with garbage results but add the known good parameters in a comment. Something like:
  ```python
   expert.train(100)  # Note: change to 100000 to train a working expert
   ```
   I like that this approach nudges the user to play around with the code and contribute something of their own to make it "work".
- We mark the notebook tests (and potentially some others too) as `pytest.mark.slow` and set up a separate slow tests job. This job would be [triggered manually](https://circleci.com/docs/2.0/workflows/#holding-a-workflow-for-a-manual-approval) e.g. just before merging a PR but not on every push.

File: examples/quickstart.py
Comment (AdamGleave) [reply examples/quickstart.py]: OK!

**Comment by AdamGleave** [comment]
> I am a bit stuck with the MCE example. I figured out how to make it execute in some sensible way but the results are garbage. Maybe I need a different RL algorithm for the learner? I just took PPO with the same hyper-parameters as for CartPole which is probably not the smartest solution. Do you have any hints for me?

The example as given seems to generate trajectories from a random policy, I assume that's because you couldn't get a good RL expert to train?

I think your best option will be to use `mce_partition_fh` from `mce_irl` and then turn it into a `BasePolicy` using `TabularPolicy` to get rollouts. This does soft value iteration to produce an entropy-regularized optimal policy.

You could also do hard value iteration, but (a) we don't have that implemented; (b) it'll cause MCE IRL to do some strange things (basically reward scale will keep growing each iteration to try and explain the deterministic policy under a softmax model of demonstrations).

File: examples/quickstart.py
Comment (AdamGleave) [reply examples/quickstart.py:1]: Thanks for the explanation, yeah I can see the value of keeping this.

File: examples/quickstart.py
Comment (AdamGleave) [inline examples/quickstart.py:57]: Consider `n_eval_episodes=3` to make it explicit (`3` seems quite magical otherwise); same for next usage on line 63

Comment (ernestum) [reply examples/quickstart.py]: Good point. I fixed it.

**Comment by ernestum** [comment]
> The example as given seems to generate trajectories from a random policy, I assume that's because you couldn't get a good RL expert to train?
> 
> I think your best option will be to use `mce_partition_fh` from `mce_irl` and then turn it into a `BasePolicy` using `TabularPolicy` to get rollouts. This does soft value iteration to produce an entropy-regularized optimal policy.

I just looked at the tests, where random trajectories are sampled. I assumed that MCE wants those random trajectories :-)

Thanks for the hint, I will try to go that way.

**Comment by ernestum** [comment]
> Thanks for the hint, I will try to go that way.

I added your suggestion. Still no meaningful outcome unfortunately. I guess it is a matter of the hyper-parameters. Unless you have a hint on what to tweak, I would just run a hp-sweep to find good ones.
I would however do that in another PR so we can finally close this one.

**Review by AdamGleave — APPROVED** [review]
LGTM.

Note I've modified MCE IRL and density baselines notebook; please review those before merging, and tag me for re-review if you end up making any significant modifications.

The MCE IRL notebook is working now but might benefit from some clean up. E.g. the pretty histogram you had didn't integrate directly with the new rollout code I used so I just ripped it out, but it wouldn't be hard to plumb it in.

Density baselines also working though some rough edges, see discussion in #412.

MCE IRL notebook depends on changes in #414, tests may not pass in this PR until that is merged.

**Review by ernestum — COMMENTED** [review]
LGTM and thanks for cleaning up after me!

**Pull request merged**
